Fix call crashes and termination sounds

cascheberg opened the pull request:
Fixes #4319, fixes #4184, fixes #5222, fixes #5353

moxie0 commented on src/org/thoughtcrime/redphone/RedPhoneService.java:
```diff
@@ -354,6 +354,26 @@ private void startCallCardActivity() {
+        RedPhoneService.this.terminate();
+      }
+    }, RedPhone.STANDARD_DELAY_FINISH);
+  }
```
Doesn't look like most of the callers need this to actually post delayed. I'd prefer not to pull these out into separate methods.

cascheberg replied:
I had tested the existing code but I could never hear the failure sound.
Probably because `outgoingRinger.playFailure()` is asynchronous and `terminate` will call `outgoingRinger.stop()`.

moxie0 replied:
let's have outgoingringer return a future then

cascheberg commented:
@moxie0 I updated the PR to use `Future`. Ready for review comments, but I'd like to do some more testing ;)
The `terminateBusy` is still using `postDelayed` because the sound is looping forever.
I removed several `private` methods in `OutgoingRinger` because they are not quite helpful.
And I fixed bugs in `SettableFuture`.

moxie0 commented on src/org/thoughtcrime/redphone/RedPhoneService.java:
```diff
@@ -357,6 +358,25 @@ private void startCallCardActivity() {
+      Log.w(TAG, e);
+    }
+    this.terminate();
+  }
```
I'd still prefer not to have dedicated methods for this

moxie0 commented on src/org/thoughtcrime/redphone/audio/OutgoingRinger.java:
```diff
@@ -37,103 +39,90 @@
+  private MediaPlayer    mediaPlayer;
+  private int            currentSoundID;
+  private Context        context;
+  private SettableFuture future;
```
I don't think this should be an instance variable, there's no need to maintain its state at the instance level and I don't see anything in here that would prevent you from calling play twice on the same instance.

cascheberg replied:
- I need to return the `Future` in `playFailure`, but I can only set the result in `onCompletion`. How else could I do that?
- One instance of `OutgoingRinger` is used to play multiple sounds, but `shutdown` handles the life cycle of the `MediaPlayer` and `SettableFuture`. Is there a problem with that?

moxie0 commented on src/org/thoughtcrime/redphone/audio/OutgoingRinger.java:
```diff
@@ -37,103 +39,90 @@
   }
 
-  private void setSound( int soundID ) {
+  private Future play(int soundID, boolean loop) {
```
always return ListenableFuture

cascheberg replied:
OK, so then I should add a `ListenableFuture.Listener` instead of calling `.get()`? That would also work around the `SettableFuture` bug.

moxie0 replied:
If you're just going to call `get` then there's not really any reason to use a future, since it can just be a blocking call. I'd use the listener since it's async now.

moxie0 commented on src/org/thoughtcrime/redphone/audio/OutgoingRinger.java:
```diff
@@ -37,103 +39,90 @@
+    } catch (IllegalArgumentException | SecurityException | IllegalStateException | IOException e) {
       Log.w(TAG, e);
-      return;
+      future.setException(e);
```
we don't propagate this exception in the synchronous case, so there's no reason to propagate it in the async case

moxie0 commented on src/org/thoughtcrime/securesms/util/concurrent/SettableFuture.java:
```diff
@@ -97,6 +99,7 @@ private void notifyAllListeners() {
     synchronized (this) {
+      notifyAll();
       localListeners = new LinkedList<>(listeners);
     }
```
It's very unlikely that I will merge changes to SettableFuture

cascheberg replied:
- The `SettableFuture.get()` is not working because it will just `wait` forever since it requires a `notify` to wake up. Fixing that is really relevant. This change adds the `notify` on all status changes, but it's ok if you have other ideas.
- The other fixes are additionally required for `cancel` to work.

moxie0 commented:
I think this is on the right track, but all in all I'd like the diff to be much simpler

cascheberg commented:
@moxie0 PR updated. I divided the patch into 10 simple commits. The first ones simplify the `OutgoingRinger` to to have only one `start` and `stop`  method instead of two each. Please have a look at them separately, I made them really small, and the existing code is really not great.
The code now uses the `ListenableFuture.Listener`.
The fix for `SettableFuture` is not required anymore, so it is not included.
All tested.

cascheberg commented:
I found a much simpler solution: only explicitly stop `OutgoingRinger` when it is looping. `playFailure()` is not looping and can take care of the shutdown on its own.

cascheberg commented:
Rebased and tested again. This PR is now a bit simpler because of other changes that have been merged in the meantime.

cascheberg commented:
rebased again

moxie0 commented:
it seems like this PR is trying to do a few different things, and the diff is too large for me right now. I'd prefer to see something that just fixes the crashes.

cascheberg commented:
I opened two new separate PRs which do not include the refactoring.
